File: Data Structures & Algorithms/4sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> sol;
        for(int i=0;i<nums.size();i++){
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for(int j=i+1;j<nums.size()-2;j++){
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                long long  target_=(long long )target-nums[i]-nums[j];
                int l=j+1;
                int r=nums.size()-1;
                if( nums[l]>0 && target_<0 ){
                    continue;
                }
                while(l<r){
                    int sum=nums[l]+nums[r];
                    if (sum==target_){
                        sol.push_back({nums[i],nums[j],nums[l],nums[r]});
                        do{
                            l++;
                        }while(nums[l]==nums[l-1]);
                        do{
                        r--;
                        }while(nums[r]==nums[r+1]);
                    }else if (sum<target_){
                        do{
                            l++;
                        }while(nums[l]==nums[l-1]);
                    }else {
                        do{
                        r--;
                        }while(nums[r]==nums[r+1]);
                    }
                }
            }
        }
        return sol;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        temp=[]
        sol=[]
        sum=0
        def rec(i:int,temp:[],sol:[],nums:List[int],target:int,sum:int):
            if (sum==target):
                sol.append(temp.copy())
                return 
            
            if (i==len(nums) or sum>target):
                return 
            
            sum=sum+nums[i]
            temp.append(nums[i])
            rec(i,temp,sol,nums,target,sum)
            sum=sum-nums[i]
            temp.pop()
            rec(i+1,temp,sol,nums,target,sum)

        rec(0,temp,sol,nums,target,sum)
        return sol
